lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -734,46 +783,189 @@ void LowerLayersPass::preprocessLayers(CircuitNamespace &ns, LayerOp layer,
+      b.create<sv::IncludeOp>(loc, IncludeStyle::Local, target);
+      break;
+    }
+
+    // If the parent layer is inline, we enable it by defining a macro.
+    // We will also have to manually enable any subsequent ancestors.
+    auto parentMacroSymbol = macroNames[parent];
+    b.create<sv::MacroDefOp>(loc, parentMacroSymbol,
+                             StringAttr::get(&getContext()));
+    parent = layer->getParentOfType<LayerOp>();
+  }
+
+  // Create IR to include bind files for child modules.
+  for (auto *record : *node) {
+    auto *child = record->getTarget()->getModule().getOperation();
+    auto files = bindFiles[child];
+    auto lookup = files.find(layer);
+    if (lookup != files.end())
+      b.create<sv::IncludeOp>(loc, IncludeStyle::Local, lookup->second.name);
+  }
+
+  // Save the bind file information for later.
+  auto &info = bindFiles[op][layer];
+  info.name = bindFileNameAttr;
+  info.body = guard.getElseBlock();
+}
+
+void LowerLayersPass::preprocessModule(CircuitNamespace &ns,
+                                       InstanceGraphNode *node) {
+  auto *op = node->getModule().getOperation();
+  if (!op)
+    return;
+
+  auto module = dyn_cast<FModuleOp>(op);
+  if (!module)
+    return;
+
+  OpBuilder b(&getContext());
+  b.setInsertionPointAfter(module);
+
+  // Create a bind file only if the layer is used under the module.
+  llvm::SmallDenseSet<LayerOp> todo;
+
+  // If the module is public, create a bind file for all layers.
+  if (module.isPublic())
+    for (auto [_, layer] : symbolToLayer)
+      if (layer.getConvention() == LayerConvention::Bind)
+        todo.insert(layer);
+
+  // Handle layers used directly in this module.
+  module->walk([&](LayerBlockOp layerBlock) {
+    auto layer = symbolToLayer[layerBlock.getLayerNameAttr()];
+    if (layer.getConvention() == LayerConvention::Inline)
+      return;
+
+    // Create a bindfile for any layer directly used in the module.
+    todo.insert(layer);
+
+    // Determine names for all modules that will be created.
+    auto moduleName = module.getModuleName();
+    auto layerName = layerBlock.getLayerName();
+
+    // A name hint for the module created from this layerblock.
+    auto layerBlockModuleName = moduleNameForLayer(moduleName, layerName);
+
+    // A name hint for the hier-path-op which targets the bound-in instance of
+    // the module created from this layerblock.
+    auto layerBlockHierPathName = hierPathNameForLayer(moduleName, layerName);
+
+    LayerBlockGlobals globals;
+    globals.moduleName = ns.newName(layerBlockModuleName);
+    globals.hierPathName = ns.newName(layerBlockHierPathName);
+    layerBlockGlobals.insert({layerBlock, globals});
+  });
+
+  // Create a bindfile for layers used indirectly under this module.
+  for (auto *record : *node) {
+    auto *child = record->getTarget()->getModule().getOperation();
+    for (auto [layer, _] : bindFiles[child])
+      todo.insert(layer);
+  }
```

Comment:
Nice use of post-order and caching the bind files in the child modules to avoid exponential behavior. 💯